src/seq/mod.rs

```diff
@@ -188,20 +188,57 @@ pub trait IteratorRandom: Iterator + Sized {
     fn choose<R>(mut self, rng: &mut R) -> Option<Self::Item>
         where R: Rng + ?Sized
     {
-        if let Some(elem) = self.next() {
-            let mut result = elem;
-            
-            // Continue until the iterator is exhausted
-            for (i, elem) in self.enumerate() {
-                let denom = (i + 2) as f64; // accurate to 2^53 elements
+        let (mut lower, mut upper) = self.size_hint();
+        let mut consumed = 0;
+        let mut result = None;
+
+        if upper == Some(lower) {
+            // Remove this once we can specialize on ExactSizeIterator
+            return if lower == 0 { None } else { self.nth(rng.gen_range(0, lower)) };
```

Comment:
Yeah, we can certainly go either way. The code *should* still compile to something quite similar since the `upper == Some(lower)` check further down is still there. But we do a few branches which the compiler likely won't be able to optimize away before getting there.

Reply:
It's seems there's a big difference when this is removed — around 6ms vs 3.7ms on the fully hinted `seq_iter_choose_from_1000` test — so worth keeping I think.